baseplate/core.py

```diff
@@ -263,6 +324,231 @@ def __init__(self):
+Session = collections.namedtuple("Session", ["id"])
+
+
+class User(_User):
+    """Wrapper for the user values in AuthenticationContext and the LoId cookie.
+    """
+
+    def event_fields(self):
+        """Dictionary of values to be added to events in the current context
+        """
+        if self.is_logged_in:
+            user_id = self.id
+        else:
+            user_id = self.loid
+        return {
+            "user_id": user_id,
+            "user_logged_in": self.is_logged_in,
+            "cookie_created": self.cookie_created_ms,
+        }
+
+    @property
+    def id(self):
+        """Authenticated account_id for the current User
+
+        :type: account_id string or None if context authentication is invalid
+        :raises: :py:class:`UndefinedSecretsException` if the
+            :py:class:`SecretsStore` has not been bound to the context handling
+            class (means that the authentication payload could not be decrypted
+            and the user_id returned)
+        :raises: :py:class:`WithheldAuthenticationError` if there was no
+            authentication token defined for the current context
+
+        """
+        return self.authentication_context.account_id
+
+    @property
+    def is_logged_in(self):
+        """Does the User have a valid, authenticated id"""
+        try:
+            return self.id is not None
+        except (WithheldAuthenticationError, UndefinedSecretsException):
+            return False
+
+    @property
+    def roles(self):
+        """Authenticated roles for the current User
+
+        :type: set(string)
+        :raises: :py:class:`UndefinedSecretsException` if the
+            :py:class:`SecretsStore` has not been bound to the context handling
+            class (means that the authentication payload could not be decrypted
+            and the user_id returned)
+        :raises: :py:class:`WithheldAuthenticationError` if there was no
+            authentication token defined for the current context
+
+        """
+        return self.authentication_context.user_roles
+
+
+class OAuthClient(_OAuthClient):
+    """Wrapper for the OAuth2 client values in AuthenticationContext."""
+
+    @property
+    def id(self):
+        """Authenticated id for the current client
+
+        :type: string or None if context authentication is invalid
+        :raises: :py:class:`UndefinedSecretsException` if the
+            :py:class:`SecretsStore` has not been bound to the context handling
+            class (means that the authentication payload could not be decrypted
+            and the user_id returned)
+        :raises: :py:class:`WithheldAuthenticationError` if there was no
+            authentication token defined for the current context
+
+        """
+        return self.authentication_context.oauth_client_id
+
+    @property
+    def type(self):
+        """Authenticated client type for the current client
+
+        :type: string or None if context authentication is invalid
```

Comment:
Because this *should* be one of a limited group of values, would it be better to translate the return value to some enum?

Reply:
I think we want this to be strings rather than Enums, I also think that the list of values should just live where the token is generated rather than in Baseplate so we can add new values without having to update Baseplate to support them.

Reply:
We just need to be _very_ careful. A similar situation in r2 was "subreddit type". At first it was just public, then public/private, then public/private/restricted, then public/private/restricted/gold, then public/private/restricted/gold/employee. Lots of checks just did stuff like "if not private" and assumed public. Is there any way we can discourage that sort of future-unproofed check?

Reply:
I don't think we can completely _stop_ that sort of thing but we could consider giving this interface rather that just returning the type. 

```python
def is_type(self, *client_types):
    return self.auth_context.oauth_client_type in client_types
```